File: C++/Arrays&Hashing/TopKFrequentElements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> count;
        vector<vector<int>> freq(nums.size() + 1, vector<int>());
        
        for(int n : nums) {
            count[n]++;
        }
        
        for(auto p : count) {
            freq[p.second].push_back(p.first);
        }
        
        vector<int> res;
        for(int i = nums.size(); i > 0; i--) {
            for(int n : freq[i]) {
                res.push_back(n);
                if(res.size() == k) {
                    return res;
                }
            }
        }
        return res;
    }
};

File: JavaScript/Arrays&Hashing/TopKFrequentElements.js
/** Solution 1
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}

var topKFrequent = function(nums, k) {
    const numCounts = {};
    for (const num of nums) {
        numCounts[num] = numCounts[num] + 1 || 1;
    }
    const sortedCounts = Object.entries(numCounts).sort((a, b) => b[1] - a[1]);
    const topK = sortedCounts.slice(0, k);
    return topK.map(([num]) => parseInt(num));
};
*/
// Solution 2 - Solution 1 condendsed
var topKFrequent = function(nums, k) {
    const numCounts = {};
    for (const num of nums) numCounts[num] = numCounts[num] + 1 || 1;
    return Object.entries(numCounts).sort((a, b) => b[1] - a[1]).slice(0, k).map(([num]) => parseInt(num));
};
